### PR TITLE
Enhanced support for limits and timeouts.

### PR DESCRIPTION
**Pull request by klizhentas**
- Add support for request size limits
- Add support for bandwidth limits
- Expose keep-alive parameters
- Move to 1.3

### CODE REVIEW

File: limit/limiter.go [hunk 1]
```diff
@@ -16,50 +16,104 @@ type Limiter interface {
...
-	Middleware
+	middleware.Middleware
 }
 
-// Mapper function takes the request and returns token that corresponds to this request
-// and the amount of tokens this request is going to consume, e.g.
+// MapperFn takes the request and returns token that corresponds to the  request and the amount of tokens this request is going to consume, e.g.
 // * Client ip rate limiter - token is a client ip, amount is 1 request
-// * Client ip memory limiter - token is a client ip, amount is number of bytes to consume
+// * Client ip bandwitdh limiter - token is a client ip, amount is number of bytes to consume
```
Comment (bradgignac) [inline limit/limiter.go]: Nitpick: bandwidth not bandwitdh

File: location/httploc/httploc.go [hunk 2]
```diff
@@ -135,16 +177,26 @@ func (l *HttpLocation) GetId() string {
...
 
 // Unwind middlewares iterator in reverse order
-func (l *HttpLocation) unwindIter(it *MiddlewareIter, req Request, a Attempt) {
+func (l *HttpLocation) unwindIter(it *middleware.MiddlewareIter, req request.Request, a request.Attempt) {
 	for v := it.Prev(); v != nil; v = it.Prev() {
 		v.ProcessResponse(req, a)
 	}
 }
 
+func (l *HttpLocation) limitRequestSize(req request.Request) (*http.Response, error) {
```
Comment (bradgignac) [inline location/httploc/httploc.go]: It looks like this method doesn't need to return a response at all.

File: location/httploc/httploc.go [hunk 3]
```diff
@@ -86,17 +116,29 @@ func NewLocationWithOptions(id string, loadBalancer LoadBalancer, o Options) (*H
...
+func (l *HttpLocation) GetObserverChain() *middleware.ObserverChain {
 	return l.observerChain
 }
 
 // Round trips the request to one of the endpoints and returns the response
-func (l *HttpLocation) RoundTrip(req Request) (*http.Response, error) {
+func (l *HttpLocation) RoundTrip(req request.Request) (*http.Response, error) {
 	originalRequest := req.GetHttpRequest()
+
+	if re, err := l.limitRequestSize(req); err != nil || err != nil {
```
Comment (bradgignac) [inline location/httploc/httploc.go]: You're duplicating your error condition on either side of the `||`.

Comment (bradgignac) [inline location/httploc/httploc.go]: What about changing this call to look like this:

```
if l.isRequestOverLimit(req) {
    return nil, errors.FromStatus(http.StatusRequestEntityTooLarge)
}

...

func (l*HttpLocation) isRequestOverLimit(req request.Request) {
    if l.options.Limits.MaxBodyBytes <= 0 {
        return false
    }
    if (req.GetHttpRequest().ContentLength < l.options.Limit.MaxBodyBytes) {
        return false
    }
    return true
}
```

**Comment by bradgignac** [comment]
Left a few comments and suggestions. Still getting familiar with the codebase, so they might not make any sense at all.

**Comment by klizhentas** [comment]
@bradgignac Thanks!

**Pull request merged**